### DISCUSSION TITLE
Expand EM approach to more ML algorithms?

### DISCUSSION THREAD

**Discussion by mxkus** [moj-analytical-services/splink#1310] [Ideas]
Hey there.

I was curious with regard to the naive bayes EM approach whether this is just a self-supervised learning approach which also could be implemented iteratively with any other algorithm like decision trees (E: train the tree with labels, M: reassign labels based on the predictions). Have I overlooked something?

If yes, there is are also fine-tunings to this approach where the retraining is only done on quite "certain" predictions (e.g. probability > 0.9). What do you think about this?

Best,
Max

**Comment by RobinL** [comment]
Hi Max - great question!
  
I'll preface this by saying I'm not an expert in self-supervised ML approaches outside those use by Splink - so take what I say with a grain of salt!

With that said, I have always wondered whether there may be some mileage in applying alternative ML techniques to the results of Splink.

A while back I tried to run a basic logistic regression using the Splink comparison vector (comparison level values) as features, and the splink prediction as the 'label'.  The idea was that it may be a way around the the 'conditional independence' assumption which is usually violated in practice (specifically conditional independence of comparison columns on match status).  The idea being that multiple regression techniques allow for correlations between features.

I didn't try that hard, but my memory is that logistic regression didn't help much*.  However:

1. I didn't try that hard
2. I didn't at the time have a very good labelled dataset to test properly whether it had worked well (e.g. AOC, ROC etc.)
3. I didn't try to use it iteratively in the way you suggest, it was just a one shot

I like the idea of using a ML model (e.g. decision trees) iteratively - it had never occurred to me to try and it's a good idea.  It certainly sounds worth experimenting e.g. with only the p > 0.9 results.  I guess whatever works best!

Finally, the biggest challenge we've faced with this sort of thing is producing robust evidence that it works, due to the absence of high quality labelled data to measure accuracy.  We do have a couple of synthetic datasets relating to persons derived from wikidata that I could potentially share if they may be of use

*  If I remember correctly - and bear in mind it was a long time ago - there was an oddity with logistic regression that the naive bayes parameter estimates essentially 'force' the logistic regression to give the same answer. Maybe this makes sense because naive literally calculates the match_probability in a kind of 'ceteris parabus' way - e.g. the effect of a match on first_name is constant irrespective of the values taken by the other parameters. Or maybe I'm talking rubbish. I didn't understand it at the time, but I probably needed to include interaction effects in the regression (which I did not do)

**Reply by RobinL** [reply]
One other thought:

I'm not sure about convergence.  With the naive bayes (Fellegi Sunter) approach, there is a proof that the log likelihood will always improve each iteration (strictly speaking, never worsen).  (see [this paper](https://imai.fas.harvard.edu/research/linkage.html)).  That generally means 

1. convergence occursand 
2. empirically speaking, convergence is usually fairly fast.  And generally the change in log likelihood is smaller each iteration.

I'm not sure whether the same would be the case with other ML models.  Not saying they wouldn't 'converge', I just don't know (and I guess the definition of convergence is not always as straightforward as in the naive bayes case where you have a likelihood function)

**Reply by zmbc** [reply]
My interpretation of your experiment results (with the caveat that you haven't described them in great detail) is that you ran a logistic regression without interaction terms and found it was nearly equivalent to F-S. I think that is expected -- F-S is known to be very closely related to *log-linear* regression, and that has been used instead of Naive Bayes. See for example [Winkler 1993](https://www.census.gov/content/dam/Census/library/working-papers/1993/adrm/rr93-12.pdf) and [Larsen and Rubin 2001](https://www.jstor.org/stable/2670338?seq=1). See the latter for some statements about convergence.

I think this is more well-trodden ground than you realized, and well worth putting into Splink without caveats about it being experimental. The ability to include interaction terms is where the approach really shines, which I am guessing you did not do, or did not do on an example that really needed it. I've experienced substantial CI issues in linking pseudopeople data with parsed addresses, and could potentially put together an example where it really helps if you could share code you used for this!

Lastly, I wanted to call your attention to more recent work from the computer science literature called [ZeroER](https://arxiv.org/pdf/1908.06049). It is also an extension of F-S that includes feature dependence, though I haven't yet exactly worked out how it relates to the log-linear approach. There is a Python implementation [here](https://github.com/chu-data-lab/zeroer).

**Reply by zmbc** [reply]
Thinking about this again, and my comment above wasn't quite right.

Machine learning algorithms that are purely [_discriminative_](https://ai.stanford.edu/~ang/papers/nips01-discriminativegenerative.pdf), such as logistic regression, are *incompatible* with EM. To spell it out: algorithms that predict the match probability given the comparison vector, **but cannot predict the comparison vector given the match probability,** are incompatible with EM.

When we say -- as I said above -- that F-S is very closely related to log-linear regression, we mean a log-linear regression predicting **the count of records given the comparison vector and match status** (with, I believe, saturated interactions between match status and everything else).

The reason logistic regression fit to the F-S/Naive Bayes result replicated the F-S/Naive Bayes result is that [they are the same model](https://en.wikipedia.org/wiki/Naive_Bayes_classifier#Relation_to_logistic_regression) when going from comparison vector -> match probability. Naive Bayes contains within it a model for predicting from match probability -> comparison vector, while logistic regression does not.

The only expansions of interest would be to go from F-S/Naive Bayes to a less naive _generative_ classifier, such as a Bayesian network. I have a hunch that some such classifier would be exactly equivalent to the log-linear regression on the counts I described above, but can't immediately work it out.

**Comment by medwar99** [comment]
I find there's significant overlap with how we do iterative improvement of our models and the Active Learning / Semi-supervised model training approaches in ML. This also helps to alleviate the labeling issue somewhat, as you build up a (partially) labelled dataset over time. There's nice techniques in the field that help visualise similarity and can help users who interact with the system. Combining this with graph-based community-finding approaches may be an interesting route.

**Reply by zmbc** [reply]
I think this is orthogonal to using more ML models for the linkage itself, because active learning [can be used (and in fact is quite effective) for the standard Fellegi-Sunter model](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3257638). I agree that this has the potential to massively improve linkages and would love to see it in Splink one day.

Graph-based community finding seems like it would be more relevant to the _clustering_ step, after performing pairwise link prediction. Right now the only clustering algorithm in Splink is connected components, but see discussions of other algorithms at #251, #2562.

**Reply by medwar99** [reply]
Agree the original question was on the model themselves, I was just thinking more holistically as an end-to-end system. The model used within an active learning setting can vary, as can methods used to sample the space during user labelling (for example generating new samples within an uncertainty region rather than just sampling from originally observed ones). Thanks for the link to Enamorado's paper, I've not read into the use of AL within the specific domain of linkage but from previous work on generic AL it makes perfect sense :)

Also agree the use of community finding in graphs is predominantly the clustering stage, but there are methods in the field which look at edge refinement as part of that strategy and blurs the line between having distinct phases for edge prediction and cluster identification (although still iterative, it isn't quite as blunt). Most of the ones I'm familiar with tend to be working on continuous spaces though, and so I'm not too up on how they'd handle a setting like this where we have discrete gammas/comparison levels as our feature vector.

**Reply by zmbc** [reply]
> the use of community finding in graphs is predominantly the clustering stage, but there are methods in the field which look at edge refinement as part of that strategy and blurs the line between having distinct phases for edge prediction and cluster identification (although still iterative, it isn't quite as blunt)

I'd be curious to learn more about this if you have a citation!

**Reply by medwar99** [reply]
[https://arxiv.org/html/2309.11798v4](https://arxiv.org/html/2309.11798v4) is a pretty thorough review, and groups the types of approach nicely. Approaches like Louvain and Label Propagation (and their successors) classically work well in graph literature, and spectral clustering/cuts too. I'm not too sure on their application within record linkage, especially with the comparison approaches we're typically using here, but I imagine there are way to use text embedding methods to put out features into a nice continuous space on which we can construct a simple distance-based graph (the problem then may be the interpretability of such an approach may not be too palatable)
